### PR TITLE
Gevent refactoring

### LINKED ISSUE #858: Blocking island server
## Describe the bug
Currently island can only handle one request at a time, so if request processing takes a while (complex report/waiting for timeout) island freezes and both, users and monkeys have to wait. Related: #857 

## To Reproduce
1. Add a sleep on version update resource, for example
2. Refresh the page. While version update resource is sleeping, island is not responsive.

## Expected behavior
We should refactor to a non-blocking WSGI container as suggested [here ](https://flask.palletsprojects.com/en/1.1.x/deploying/wsgi-standalone/#tornado)

### PR DESCRIPTION
**Pull request by VakarisZ**
# What does this PR do? 

Fixes #858 #857

Refactored tornado WSGI container into gevent WSGI container.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing? 

## Testing Checklist

* [ ] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested locally, on windows

### CODE REVIEW

**Comment by acepace** [comment]
While i take a look
Can you explain the need? I can guess but want to make sure.

**Comment by VakarisZ** [comment]
@acepace take a look into the issues this fixes.

File: monkey/infection_monkey/control.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def find_server(default_tunnel=None):
                 if ControlClient.proxies:
                     debug_message += " through proxies: %s" % ControlClient.proxies
                 LOG.debug(debug_message)
-                requests.get("https://%s/api?action=is-up" % (server,),  # noqa: DUO123
+                requests.get(f"https://{server}/api?action=is-up",  # noqa: DUO123
```
Comment (acepace) [inline monkey/infection_monkey/control.py:84]: What version of python are we baselined on? 3.6 or 3.7?

File: monkey/infection_monkey/post_breach/job_scheduling/windows_job_scheduling.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 
 def get_windows_commands_to_schedule_jobs():
-    return f'schtasks /Create /SC monthly /TN {SCHEDULED_TASK_NAME} /TR {SCHEDULED_TASK_COMMAND}'
+    return f'schtasks /Create /SC monthly /F /TN {SCHEDULED_TASK_NAME} /TR {SCHEDULED_TASK_COMMAND}'
```
Comment (acepace) [inline monkey/infection_monkey/post_breach/job_scheduling/windows_job_scheduling.py:8]: Shouldn't be part of this PR but shrug

File: monkey/infection_monkey/post_breach/actions/schedule_jobs.py [hunk 3]
```diff
@@ -15,5 +15,7 @@ def __init__(self):
         super(ScheduleJobs, self).__init__(name=POST_BREACH_JOB_SCHEDULING,
                                            linux_cmd=' '.join(linux_cmds),
                                            windows_cmd=windows_cmds)
-
+
```
Comment (acepace) [inline monkey/infection_monkey/post_breach/actions/schedule_jobs.py:18]: Shouldn't be part of this PR but shrug

File: monkey/infection_monkey/post_breach/post_breach_handler.py [hunk 4]
```diff
@@ -25,9 +25,9 @@ def execute_all_configured(self):
         """
         Executes all post breach actions.
         """
-        pool = Pool(5)
-        pool.map(self.run_pba, self.pba_list)
-        LOG.info("All PBAs executed. Total {} executed.".format(len(self.pba_list)))
+        with Pool(5) as pool:
+            pool.map(self.run_pba, self.pba_list)
+            LOG.info("All PBAs executed. Total {} executed.".format(len(self.pba_list)))
```
Comment (acepace) [inline monkey/infection_monkey/post_breach/post_breach_handler.py:30]: Shouldn't be part of this PR but shrug

File: monkey/infection_monkey/post_breach/post_breach_handler.py [hunk 5]
```diff
@@ -40,5 +40,6 @@ def run_pba(self, pba):
         try:
             LOG.debug("Executing PBA: '{}'".format(pba.name))
             pba.run()
+            LOG.debug(f"Execution of {pba.name} finished")
```
Comment (acepace) [inline monkey/infection_monkey/post_breach/post_breach_handler.py:43]: Shouldn't be part of this PR but shrug

File: monkey/monkey_island.py [hunk 6]
```diff
@@ -1,3 +1,6 @@
+from gevent import monkey
+monkey.patch_all()
```
Comment (acepace) [inline monkey/monkey_island.py:2]: Confusing name, wouldn't it be nicer to
from gevent import monkey as gevent_monkey

since our entire codebase uses monkey a lot

File: monkey/monkey_island/cc/services/infection_lifecycle.py [hunk 7]
```diff
@@ -8,8 +8,8 @@
 from monkey_island.cc.services.config import ConfigService
 from monkey_island.cc.services.node import NodeService
 from monkey_island.cc.services.reporting.report import ReportService
-from monkey_island.cc.services.reporting.report_generation_synchronisation import (
-    is_report_being_generated, safe_generate_reports)
+from monkey_island.cc.services.reporting.report_generation_synchronisation import safe_generate_reports, \
+    is_report_being_generated
```
Comment (acepace) [inline monkey/monkey_island/cc/services/infection_lifecycle.py:12]: Shouldn't be part of this PR but shrug

File: monkey/monkey_island/cc/services/reporting/report_generation_synchronisation.py [hunk 8]
```diff
@@ -1,15 +1,15 @@
 import logging
-import threading
+from gevent.lock import Semaphore
```
Comment (acepace) [inline monkey/monkey_island/cc/services/reporting/report_generation_synchronisation.py:2]: BoundedSemaphore

**Comment by acepace** [comment]
LGTM

File: monkey/infection_monkey/control.py (hunk 1 above)
Comment (VakarisZ) [reply monkey/infection_monkey/control.py:84]: 3.7

File: monkey/infection_monkey/post_breach/job_scheduling/windows_job_scheduling.py (hunk 2 above)
Comment (VakarisZ) [reply monkey/infection_monkey/post_breach/job_scheduling/windows_job_scheduling.py]: Fixed, didn't notice other commits

**Review by acepace — APPROVED** [review]

**Comment by acepace** [comment]
Can you rebase and we merge?

**Review by itaymmguardicore — COMMENTED** [review]
Overall looks good to me. If @acepace doesn't have any further comments I'm good with it.
My only concern is the transition from single-thread to multi-thread can cause unexpected errors in different parts in code which weren't designed properly or didn't take this into consideration.

I suggest running some/all automatic blackbox tests on this branch to make sure we didn't break anything

**Comment by VakarisZ** [comment]
@itaymmguardicore so let's do a proper tests before the next release and catch all errors, not only something caused by gevent. Cost of running in depth tests > info that specific bug happened due to gevent.
